### ISSUE TITLE
Producer.Supervisor crashes if topic does not exist

### ISSUE THREAD

**Issue by keathley** [bbalser/elsa#58]
Hey y'all,

I ran into a scenario today where I tried to start a producer with a topic that hadn't been created yet. If you do that it'll cause a crash:

```elixir
** (Mix) Could not start application my_consumer: MyConsumer.Application.start(:normal, []) returned an error: shutdown: failed to start child: MyConsumer.Consumer
    ** (EXIT) shutdown: failed to start child: :producer_supervisor_some_topic
        ** (EXIT) an exception was raised:
            ** (MatchError) no match of right hand side value: {:error, :leader_not_available}
                (elsa) lib/elsa/producer/supervisor.ex:34: Elsa.Producer.Supervisor.init/1
                (stdlib) supervisor.erl:295: :supervisor.init/1
                (stdlib) gen_server.erl:374: :gen_server.init_it/2
                (stdlib) gen_server.erl:342: :gen_server.init_it/6
                (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

The crash is caused here: https://github.com/bbalser/elsa/blob/master/lib/elsa/producer/supervisor.ex#L34. In this scenario, we don't match on `{:ok, partitions}` and this causes the supervisor to crash.

I'm not sure if this is intended behaviour or not? I see arguments either way. If we allowed the producer to start with an unknown number of partitions it would push the error handling logic into the producer which may not be desirable. It also implies that we'd need to do some sort of polling to see what the true number of partitions is for a topic. This may not be a terrible idea since the number of partitions could change dynamically. Polling would allow for this dynamism.

I wanted to get y'all's thoughts on it. I'm happy to work on a fix if you feel like this is "broken". But if this is working as intended I'm happy to find another way around the problem.

**Comment by jeffgrunewald** [comment]
Hey, thanks reaching out. I'd say the behavior is partially intended, albeit unintended.

Crashing is certainly not the desired behavior, but we already have a solution in mind in response to github.com/bbalser/elsa/issues/55 that we would apply to the producer unable to find a topic at all. We hope to start pushing out a working branch this week and get it merged shortly after that. That's going to come with some changes to the supervision strategy that should actually make dynamic polling of the topic for changes to the number of partitions pretty easy.

Solving 55 will, in the short term, push a consumer or producer supervision tree into a configurable retry loop before starting to prevent the crashing. After that fix is released adding the ability to respond to a change on the topic feels like a good feature to add.

**Comment by keathley** [comment]
That sounds great. Let me know if I can help out

**Comment by jeffgrunewald** [comment]
Just checking in that this effort is almost done; we've been pretty busy pushing toward something only loosely related at work, but this should be merged in soon.

**Comment by keathley** [comment]
@jeffgrunewald No rush dawg :). Let me know if I can do anything to help out.

**Comment by jeffgrunewald** [comment]
FYI, @bbalser just released version 0.12 that _should_ address this issue and takes into account reactive responsiveness to changes on the consumer side while relying on polling to detect changes from a producer perspective. Let us know if there's anything that you think we missed!
